core/shared/src/main/scala/scalaz/zio/Promise.scala

```diff
@@ -112,6 +112,26 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
       action
     })
 
+  private[zio] final def unsafeDone(r: ExitResult[E, A]): Unit = {
+    var retry: Boolean                = true
+    var joiners: List[Callback[E, A]] = null
+
+    while (retry) {
+      val oldState = state.get
+
+      val newState = oldState match {
+        case Pending(js) =>
+          joiners = js
+          Done(r)
+        case _ => oldState
+      }
+
+      retry = state.compareAndSet(oldState, newState)
+    }
+
+    if (joiners ne null) joiners.foreach(_(r))
```

Comment:
This one is going to be the problem... hmm. We need to do this on the thread pool (like the `forkAll_` implementation), since each joiner could take infinitely long time to return.

Reply:
Ah yeah, I see. It looks like a dealbreaker.

Reply:
Ah, nope, we can do it!

```scala
def unsafeDone(exit: ExitResult[E, A], executor: Runnable => Unit): Unit = ???
```